### ISSUE TITLE
realm: gasnetex network layer (documentation/tracking issue)

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#986]
NOTE: Original issue was #628 which may still be of interest to folks.  This issue is acting as initial documentation and a centralized location for reporting of issues and/or results.

What: The gasnetex network layer is a ground up rewrite of Realm's use of GASNet for inter-node communication, taking advantage of a number of features of GASNet-EX that were not present in GASNet-1.

Requirements:
1. The GASNet source, version 2020.11.0 or newer, obtainable from https://gasnet.lbl.gov
2. For dynamic registration of system memory, an Infiniband network, and all the right drivers/modules/etc.
3. For GPUDirect goodness, the above and NVIDIA Tesla/Quadro GPU(s)
4. An configuration of GASNet appropriate for both your system and for Legion.  Examples can be found here: https://github.com/StanfordLegion/gasnet/tree/master/configs
5. (for the moment) Either the `gex` or `gex_cr` branch of Legion

Building Legion with the gasnetex network layer:
1. for Makefile-based builds, set `REALM_NETWORKS=gasnetex` (do NOT set the old `USE_GASNET=1` option) and `GASNET_ROOT=...`
2. for CMake-based builds, set `Legion_NETWORKS=gasnetex` (do not set `Legion_USE_GASNet`) and `-DGASNet_ROOT_DIR=...` if needed

Running your jobs is done as with the original (called `gasnet1` in our build, if you're curious) networking code, using `gasnetrun_(conduit)` or the mpi interop-assisted `mpirun`, `srun`, etc. (strongly preferred).  A number of command-line configuration variables have been added and may be used to tune your configuration and/or help narrow down the cause of any issues observed:
* `-gex:bindhost [0|1]` - if set to `1` (default), Realm will attempt to bind all system memory allocations (e.g. the stuff controlled by `-ll:csize` and `-ll:zsize`) with GASNet, allowing direct inter-node copies - this should result in significantly better bandwidth and somewhat better latency compared to the gasnet1 implementation
* `-gex:bindcuda [0|1]` - if set to `1` (default), Realm will attempt to bind CUDA device memory, using GPUDirect for inter-node copies (to device or system memory on the remote node) (IMPORTANT: on many systems, the amount of device memory that can be bound is limited, and with GASNet 2020.11.0, exceeding this limit results in a fatal error.  The next version of GASNet is expected to return an error code for Realm to deal with.)
* `-gex:immediate [0|1]` - if set to `1` (default), threads other than the GASNet progress thread (e.g. application threads or other realm workers) may attempt to inject packets into the network directly, using GASNet-EX's support for returning a result if the packet cannot be sent immediately, getting improved message latency while preventing the thread from stalling due to network resource limitations.  If congestion is detected (or if this mode is disabled), packets are queued up to be handled by the progress thread.
* `-gex:negotiated [0|1]` - if set to `1` (default), Realm uses GASNet-EX's support for "negotiated payload" active messages, avoiding a copy in the outbound path by letting Realm assembly a packet directly in GASNet-managed memory
* `-gex:batch [0|1]` - if set to `1` (default), Realm will attempt to batch up small messages on the network to improve throughput when congestion is observed.  Batching is opportunistic and should never increase message latency.
* `-gex:maxlong [bytes]` (default is `4m` - i.e. 4MB) - controls the maximum size of long (RDMA) transfer Realm will present to GASNet as a single request.  Ideally this is just large enough to get peak network bandwidth, but no larger in order to prevent enormous packets from blocking subsequent messages that might be latency-sensitive.  Tuning this is a per-system thing - you can use GASNet's `testlarge` to find the sweet spot for your system.
* `-gex:cksum [0|1]` - if set to `1` (current default, but likely to be disabled before Legion 2020.12 release), Realm computes a checksum for all outbound packets and verifies it on reception.  Detected mismatches are mostly likely to be indicative of Realm bugs, but could theoretically be GASNet or network HW/SW issues.  Consider turning this on (only) if you are getting corrupted data in copies or if the Legion runtime is getting confused.
* `-gex:obcount [count]` (default is `64`) - number of output buffers to request from GASNet at initialization time.  These buffers are used for both queued messages and for assembling of data for multi-dimensional/sparse/indirect copies.  Unlike the old gasnet1 implementation, these buffers are shared across all endpoints, greatly reducing the number required, but at least 2 per node in the job is recommended.  If these output buffers are exhausted, Realm will dynamically allocate overflow buffers, increasing memory consumption and adding overhead due to extra copies of message data.
* `-gex:obsize [bytes]` (default is `256k`) - sets the size of each output buffer.  Larger buffers reduce bookkeeping overhead and permit larger message batches, while smaller buffers reduce overall memory consumption.  (NOTE: Several places in Realm currently assume that output buffers are at least 65KB, but the intent is to find/fix those cases and get down into the 16KB range as the minimum allowed.)

**Comment by lightsighter** [comment]
Probably want to add that you want to build GASNet from the Legion gasnet repository with `ibvcuda`. That's actually better than downloading GASNet directly from their website and figuring out that you have to pass the magic `--enable-kind-cuda-uva` flag.

**Comment by streichler** [comment]
No, I need to delete the `ibvcuda` config, as it makes the name not match the conduit.  Please try to use the `ibv` config - it should auto-detect CUDA (or we should figure out why it fails to do so).

**Comment by bonachea** [comment]
@streichler @lightsighter : 

In the 2020.11.0 GASNet prototype release, the configure default is to probes for CUDA support (because there was no other reason to use the prototype).

In the forthcoming stable release, there will be a single release tarball but the GPUDirect prototype feature will default to off (so users don't get prototype features without opting-in). 

So you might want to add an explicit `--enable-kind-cuda-uva=probe` configure argument once you pull in our forthcoming release.

CC @PHHargrove
